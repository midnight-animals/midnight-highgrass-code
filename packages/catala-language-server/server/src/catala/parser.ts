import * as fs from 'fs';
import * as path from 'path';
import * as Parser from 'tree-sitter';
// import * as Catala from '/home/user/dev/projects/tree-sitter/tree-sitter-catala/bindings/node/index.js';
import * as Catala from '../catala/tree-sitter/bindings/node';
import {
  CatalaFileParsed,
  CatalaGrammarTypes,
  CatalaSyntaxNode,
} from '../shared/CatalaTypes';
import {
  getAscendantNodeTypesPath,
  getAscendantTypesUntilSourceFile,
  getClosestAscendantOfTypes,
  getTypeFromClosestAscendantOfTypes,
} from './tree-sitter-focus/tree-sitter-modules';
import {
  areArraysEqual,
  arrayAlreadyIncludesArray,
} from '../shared/arrayUtils';

export async function parse(text: string): Promise<CatalaFileParsed> {
  const parser = new Parser();
  parser.setLanguage(Catala);
  const tree = parser.parse(text);
  const catalaFile: CatalaFileParsed = {
    error: [],
    constructor_name: [],
    enum_struct_name: [],
    field_name: [],
    scope_name: [],
    variable: [],
    verb_block: [],
    tree,
  };
  walkRecursive(tree.rootNode as CatalaSyntaxNode, catalaFile);
  return catalaFile;
}

function walkRecursive(
  node: CatalaSyntaxNode | undefined,
  catalaFile: CatalaFileParsed
) {
  if (!node) return;
  const catalaSyntaxNode: CatalaSyntaxNode = Object.assign(node, {
    enclosingBlockType: 'ALL',
  });

  // ['enum_struct_name', 'scope_decl_item', 'scope_decl']
  if (node.type === 'ERROR') {
    catalaFile.error.push(catalaSyntaxNode);
  }

  if (node.type === 'constructor_name') {
    const collectedTypes: CatalaGrammarTypes[] = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    if (collectedTypes.length > 2) {
      /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
      node.type; /*?*/
      collectedTypes; /*?*/
    }

    const parentTypes: CatalaGrammarTypes[] = ['enum_decl_item'];
    const grandParentTypes: CatalaGrammarTypes[] = ['enum_decl'];
    const ascendantTypesPath = getAscendantNodeTypesPath(
      node,
      parentTypes,
      grandParentTypes
    );
    catalaSyntaxNode.ascendantTypesPath = ascendantTypesPath;
    catalaFile.constructor_name.push(catalaSyntaxNode);
  } else if (node.type === 'enum_struct_name') {
    const collectedTypes: CatalaGrammarTypes[] = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    if (collectedTypes.length > 2) {
      /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
      node.type; /*?*/
      collectedTypes; /*?*/
    }

    const parentTypes: CatalaGrammarTypes[] = [
      'definition',
      'enum_decl',
      'struct_decl',
      'scope_decl_item',
    ];
    const grandParentTypes: CatalaGrammarTypes[] = ['scope', 'scope_decl'];
    const ascendantTypesPath = getAscendantNodeTypesPath(
      node,
      parentTypes,
      grandParentTypes
    );
    catalaSyntaxNode.ascendantTypesPath = ascendantTypesPath;
    catalaFile.enum_struct_name.push(catalaSyntaxNode);
  } else if (node.type === 'field_name') {
    const collectedTypes: CatalaGrammarTypes[] = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    if (collectedTypes.length > 2) {
      /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
      node.type; /*?*/
      collectedTypes; /*?*/
    }

    const parentTypes: CatalaGrammarTypes[] = [
      'definition',
      'struct_decl',
      'struct_decl_item',
    ];
    const grandParentTypes: CatalaGrammarTypes[] = [
      'scope',
      'scope_decl',
      'struct_decl',
    ];
    const ascendantTypesPath = getAscendantNodeTypesPath(
      node,
      parentTypes,
      grandParentTypes
    );
    catalaSyntaxNode.ascendantTypesPath = ascendantTypesPath;
    catalaFile.field_name.push(catalaSyntaxNode);
  } else if (node.type === 'scope_name') {
    const collectedTypes: CatalaGrammarTypes[] = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    if (collectedTypes.length > 2) {
      /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
      node.type; /*?*/
      collectedTypes; /*?*/
    }

    const parentTypes: CatalaGrammarTypes[] = [
      'scope_decl_item',
      'scope',
      'scope_decl',
    ];
    const grandParentTypes: CatalaGrammarTypes[] = ['scope', 'scope_decl'];
    const ascendantTypesPath = getAscendantNodeTypesPath(
      node,
      parentTypes,
      grandParentTypes
    );
    catalaSyntaxNode.ascendantTypesPath = ascendantTypesPath;
    catalaFile.scope_name.push(catalaSyntaxNode);
  } else if (node.type === 'variable') {
    const collectedTypes: CatalaGrammarTypes[] = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
    node.type; /*?*/
    collectedTypes; /*?*/

    const parentTypes: CatalaGrammarTypes[] = [
      'assertion',
      'definition',
      'rule',
      'scope_decl_item',
    ];
    const grandParentTypes: CatalaGrammarTypes[] = ['scope', 'scope_decl'];
    const ascendantTypesPath = getAscendantNodeTypesPath(
      node,
      parentTypes,
      grandParentTypes
    );

    catalaSyntaxNode.ascendantTypesPath = ascendantTypesPath;
    catalaFile.variable.push(catalaSyntaxNode);
  } else if (node.type === 'verb_block') {
    const collectedTypes: CatalaGrammarTypes[] = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    if (collectedTypes.length > 2) {
      /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
      node.type; /*?*/
      collectedTypes; /*?*/
    }

    const parentTypes: CatalaGrammarTypes[] = ['definition', 'scope_decl_item'];
    const grandParentTypes: CatalaGrammarTypes[] = ['scope', 'scope_decl'];
    const ascendantTypesPath = getAscendantNodeTypesPath(
      node,
      parentTypes,
      grandParentTypes
    );

    catalaSyntaxNode.ascendantTypesPath = ascendantTypesPath;
    catalaFile.variable.push(catalaSyntaxNode);
  }

  if (node.childCount === 0) return;

  node.children.forEach((child) => {
    walkRecursive(child, catalaFile);
  });
}

const content = fs.readFileSync(
  path.resolve(__dirname, '../..', 'tests/fixtures/cli.catala_en'),
  'utf-8'
);

async function main() {
  const result = await parse(content).catch((err) => console.error(err));
  const { tree, ...parsed } = result as any;

  const map: Record<string, CatalaGrammarTypes[][]> = {};
  Object.entries(parsed).forEach(([key, value]) => {
    map[key] = [];
    parsed[key].forEach((node: CatalaSyntaxNode) => {
      if (arrayAlreadyIncludesArray(map[key], node.ascendantTypesPath)) return;

      if (node.ascendantTypesPath.length === 0) {
        node.ascendantTypesPath; /*?*/
        node.parent.parent.text; /*?*/
        node.text; /*?*/
      }

      map[key].push(node.ascendantTypesPath);
    });
  });

  map; /*?*/
  // parsed.enum_struct_name[0].ascendantTypesPath; /*?*/
}
main();
