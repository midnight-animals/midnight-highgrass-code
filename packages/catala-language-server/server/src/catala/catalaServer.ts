import {
  Connection,
  Diagnostic,
  DiagnosticSeverity,
  Position,
  TextDocuments,
} from 'vscode-languageserver';
import { TextDocument } from 'vscode-languageserver-textdocument';
import { parse } from './parser';
import { CatalaFile } from '../shared/CatalaTypes';
import { UriUtils } from '../shared/uri-utils';
import { onCompletion } from './capabilities/onCompletion';
import { onDefinition } from './capabilities/onDefinition';
import { convertNodeToRange } from '../shared/vscodeUtils';
import { closestAscendantOfType } from './tree-sitter-focus/tree-sitter-modules';
import { getCatalaCliOutput } from './modules/getCatalaCliOutput';

export class CatalaServer {
  private catalaFiles: CatalaFile[] = [];

  constructor(
    private connection: Connection,
    private documents: TextDocuments<TextDocument>
  ) {}

  public async validateTextDocument(textDocument: TextDocument) {
    console.log('[INFO] Validating text document...');
    if (this.catalaFiles.length === 0) return Promise.resolve([]);

    const targetFile = this.catalaFiles.find(
      (f) => f.uri === UriUtils.toSysPath(textDocument.uri)
    );
    if (!targetFile) return Promise.resolve([]);
    if (!targetFile.error.length) return Promise.resolve([]);

    const diagnostics: Diagnostic[] = [];

    const cliOutput = await getCatalaCliOutput(targetFile.tree.rootNode.text);
    if (!cliOutput?.range) return Promise.resolve([]);
    const message = cliOutput.text;
    console.error(message, ' at\n', cliOutput.range);
    const diagnostic: Diagnostic = {
      severity: DiagnosticSeverity.Error,
      range: cliOutput.range,
      source: 'catala-language-server',
      message,
    };
    diagnostics.push(diagnostic);

    return Promise.resolve(diagnostics);
  }

  public async onInitialized() {
    this.catalaFiles = [];
    // TODO ISSUE-JcuPKubs
    const extensionSettings = await this.connection.workspace.getConfiguration({
      section: '',
    });

    // parse current document
    for await (const document of this.documents.all()) {
      const content = document.getText();
      try {
        const parsed = await parse(content);
        const uri = UriUtils.toSysPath(document.uri);
        const file: CatalaFile = {
          ...parsed,
          uri,
        };
        this.catalaFiles.push(file);
      } catch (error) {
        console.error(error);
      }
    }

    // const logMe = JSON.stringify(this.catalaFiles[0], null, 2);
  }

  public onCompletion() {
    return onCompletion(this.catalaFiles);
  }

  public async onDefinition(document: TextDocument, position: Position) {
    const normalizedUri = UriUtils.toSysPath(document.uri);
    const targetFile = this.catalaFiles.find(
      (file) => file.uri === normalizedUri
    );
    if (!targetFile) return [];

    const definitions = onDefinition(targetFile, position);
    return definitions;
  }
}
