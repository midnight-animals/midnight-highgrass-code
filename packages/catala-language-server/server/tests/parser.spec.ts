import {
  TEST_FILE_NAMES,
  readCatalaFile,
} from './test-helpers/readCatalaTestFile';
import { parse } from '../src/catala/parser';
import { matchSnapshot } from './test-helpers/matchSnapshot';

describe('parser', () => {
  TEST_FILE_NAMES.forEach((fileName) => {
    it(fileName, async () => {
      const content = readCatalaFile(fileName);
      const parsed = await parse(content);
      matchSnapshot(content, parsed);
    });
  });
});
