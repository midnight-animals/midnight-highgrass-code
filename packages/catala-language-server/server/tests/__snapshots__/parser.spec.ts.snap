// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser combination_declStruct_declScope_scope 1`] = `
{
  "content": "
\`\`\`catala
declaration structure Individual:
  data income content money
\`\`\`
\`\`\`catala
declaration scope IncomeTaxComputation:
  input individual content Individual
\`\`\`
\`\`\`catala
scope IncomeTaxComputation:
  definition income_tax equals
    individual.income * fixed_percentage
\`\`\`
",
  "props": {
    "enum_struct_name": [
      {
        "endPosition": {
          "column": 32,
          "row": 1,
        },
        "startPosition": {
          "column": 22,
          "row": 1,
        },
        "text": "Individual",
      },
      {
        "endPosition": {
          "column": 37,
          "row": 6,
        },
        "startPosition": {
          "column": 27,
          "row": 6,
        },
        "text": "Individual",
      },
    ],
    "field_name": [
      {
        "endPosition": {
          "column": 13,
          "row": 2,
        },
        "startPosition": {
          "column": 7,
          "row": 2,
        },
        "text": "income",
      },
      {
        "endPosition": {
          "column": 21,
          "row": 11,
        },
        "startPosition": {
          "column": 15,
          "row": 11,
        },
        "text": "income",
      },
    ],
    "scope": [
      {
        "endPosition": {
          "column": 40,
          "row": 11,
        },
        "startPosition": {
          "column": 0,
          "row": 9,
        },
        "text": "scope IncomeTaxComputation:
  definition income_tax equals
    individual.income * fixed_percentage",
      },
    ],
    "variable": [
      {
        "endPosition": {
          "column": 18,
          "row": 6,
        },
        "startPosition": {
          "column": 8,
          "row": 6,
        },
        "text": "individual",
      },
      {
        "endPosition": {
          "column": 23,
          "row": 10,
        },
        "startPosition": {
          "column": 13,
          "row": 10,
        },
        "text": "income_tax",
      },
      {
        "endPosition": {
          "column": 14,
          "row": 11,
        },
        "startPosition": {
          "column": 4,
          "row": 11,
        },
        "text": "individual",
      },
      {
        "endPosition": {
          "column": 40,
          "row": 11,
        },
        "startPosition": {
          "column": 24,
          "row": 11,
        },
        "text": "fixed_percentage",
      },
    ],
  },
}
`;

exports[`parser scope_and_definition 1`] = `
{
  "content": "
\`\`\`catala
scope IncomeTaxComputation:
  definition income_tax equals
    individual.income
\`\`\`
",
  "props": {},
}
`;

exports[`parser scope_decl 1`] = `
{
  "content": "
\`\`\`catala
declaration scope IncomeTaxComputation:
  input individual content Individual
\`\`\`
",
  "props": {},
}
`;

exports[`parser struct_decl 1`] = `
{
  "content": "
\`\`\`catala
declaration structure Individual:
\`\`\`
",
  "props": {},
}
`;

exports[`parser struct_decl_with_1_field 1`] = `
{
  "content": "
\`\`\`catala
declaration structure Individual:
  data income content money
\`\`\`
",
  "props": {},
}
`;
